[Cache] Stop defaulting to `igbinary` in `DefaultMarshaller`

martijnc opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? |no
| Issues        | Fix #52391
| License       | MIT

igbinary used to be a drop in replacement for PHP's `serialize` but recent changes (in PHP 7.4) to the handling of uninitialized properties in `serialize` have not made it into igbinary, so it no longer is a simple drop in replacement.

This only removes igbinary as the default serializer, code can still opt-in through the first constructor argument.

This may result in a performance regression on systems with igbinary as it would no longer be used by default.

Fixes #52391

derrabus commented:
Changing the default behavior silently in a bugfix release on an LTS branch sounds like a bad idea, tbh.

If we want to do this, we should target 7.2 and deprecate not setting this option first.

Commit:
[Cache] Stop defaulting to igbinary in `DefaultMarshaller`

igbinary used to be a drop in replacement for PHP's `serialize` but recent changes (in PHP 7.4) to the handling of uninitialized properties in `serialize` have not made it into igbinary, so it no longer is a simple drop in replacement.

This only removes igbinary as the default serializer, code can still opt-in through the first constructor argument.

This may result in a performance regression on systems with igbinary as it would no longer be used by default.

nicolas-grekas commented:
The code is able to auto-detect if igbinary was used when serializing so that payloads are compatible with either flags (if igbinary is installed).
This means changing the default behavior is fine without further precautions.
We could even do this on 5.4 if we think the state of igbinary is too bad.

xabbuh commented:
> This means changing the default behavior is fine without further precautions.

I am not convinced that this is always true. What if other PHP projects not using Symfony read the serialised data? They may not have the same mechanism to detect if igbinary was used.

nicolas-grekas commented:
> What if other PHP projects not using Symfony read the serialised data? They may not have the same mechanism to detect if igbinary was used.

This would be an unsupported use case. Cached items are certainly not meant for interop. We already changed their format without caring in the past.

This change is desired to me.

nicolas-grekas reviewed (approved):
Works for me on 7.2. I'd also be fine on 5.4/6.4: igbinary is not safe anymore, ppl are at risk when using it, and the fact that it's not opt-in when the extension is installed adds to the risk.

nicolas-grekas commented:
See eg comments in https://github.com/igbinary/igbinary/issues/273

chalasr reviewed (approved):
👍 for 5.4

derrabus commented:
> Works for me on 7.2.

I think, 7.2 is the compromise we can all settle on. And please let's add a remark to the upgrade guide for 7.2 and the Cache component's change log.

> I'd also be fine on 5.4/6.4: igbinary is not safe anymore, ppl are at risk when using it, and the fact that it's not opt-in when the extension is installed adds to the risk.

I understand that and maybe it was even a mistake to enable igbinary serialization automatically. But I believe that there are enough projects out there that don't hit those issues and still rely on the performance gain they get from leveraging igbinary. Maybe I'm too conservative here, but I wouldn't want to lure those apps into a performance degradation with an innocent bugfix release.

xabbuh commented on src/Symfony/Component/Cache/Marshaller/DefaultMarshaller.php:
```diff
@@ -25,9 +25,8 @@ class DefaultMarshaller implements MarshallerInterface
-        if (null === $useIgbinarySerialize) {
-            $useIgbinarySerialize = \extension_loaded('igbinary') && version_compare('3.1.6', phpversion('igbinary'), '<=');
-        } elseif ($useIgbinarySerialize && (!\extension_loaded('igbinary') || version_compare('3.1.6', phpversion('igbinary'), '>'))) {
+        $useIgbinarySerialize = true === $useIgbinarySerialize;
```
I would remove this line

fabpot commented:
Thank you @martijnc.
